lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -101,23 +115,83 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
-        enable_zones = (availability_zones || []).dup
-        disable_zones = []
-        actual_elb.availability_zones.each do |availability_zone|
-          if !enable_zones.delete(availability_zone.name)
-            disable_zones << availability_zone.name
+        # TODO: refactor this whole giant method into many smaller method calls
+        # Update scheme - scheme is immutable once set, so if it is changing we need to delete the old
+        # ELB and create a new one
+        if scheme.downcase != actual_elb.scheme
+          desc = ["  updating scheme to #{scheme}"]
+          desc << "  WARN: scheme is immutable, so deleting and re-creating the ELB"
+          perform_action.call(desc) do
+            actual_elb.delete
+            actual_elb = elb.load_balancers.create(lb_spec.name, lb_optionals)
           end
         end
-        if enable_zones.size > 0
-          perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
-            actual_elb.availability_zones.enable(*enable_zones)
+
+        # Update security groups
+        if security_group_ids.empty?
+            Chef::Log.info("No Security Groups specified in resource; load_balancer[#{actual_elb.name}] cannot have " +
+            "empty Security Groups, so assuming it only currently has the default Security Group.  No action taken.")
+        else
+          current = Set.new(actual_elb.security_group_ids)
+          desired = Set.new(security_group_ids)
+          if current != desired
+            perform_action.call("  updating security groups to #{desired.to_a}") do
+              elb.client.apply_security_groups_to_load_balancer(
+                load_balancer_name: actual_elb.name,
+                security_groups: desired.to_a
+              )
+            end
           end
         end
-        if disable_zones.size > 0
-          perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
-            actual_elb.availability_zones.disable(*disable_zones)
+
+        # You can only have 1 of either subnets or availability zones, and if in a VPC you must have
+        # at least one of either.
+        if !availability_zones.empty?
+
+          enable_zones = (availability_zones || []).dup
+          disable_zones = []
+          actual_elb.availability_zones.each do |availability_zone|
+            if !enable_zones.delete(availability_zone.name)
+              disable_zones << availability_zone.name
+            end
+          end
+          if enable_zones.size > 0
+            perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
+              actual_elb.availability_zones.enable(*enable_zones)
+            end
           end
+          if disable_zones.size > 0
+            perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
+              actual_elb.availability_zones.disable(*disable_zones)
+            end
+          end
+
+        elsif !subnets.empty?
```

Comment:
What happens if the user starts out with a load balancer with subnets, and then changes it to have availability_zones?  (Or vice versa.)  It might be best to unconditionally calculate the enable/disable/attach/detach arrays, so that subnets can be removed and then availability zones added.